clippy_lints/src/duplicate_match_guards.rs

```diff
@@ -0,0 +1,171 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::{
+    HasSession, SpanRangeExt, indent_of, reindent_multiline, snippet_with_applicability, trim_span,
+};
+use clippy_utils::{eq_expr_value, span_contains_comment};
+use rustc_errors::Applicability;
+use rustc_hir::{Arm, ExprKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+use rustc_span::BytePos;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for the same condition being checked in a match guard and in the match body
+    ///
+    /// ### Why is this bad?
+    /// This is usually just a typo or a copy and paste error.
+    ///
+    /// ### Known problems
+    /// False negatives: if the condition is an impure function, it could've been called twice on
+    /// purpose for its side effects
+    ///
+    /// ### Example
+    /// ```no_run
+    /// # let n = 0;
+    /// # let a = 3;
+    /// # let b = 4;
+    /// match n {
+    ///     0 if a > b => {
+    ///         if a > b {
+    ///             return;
+    ///         }
+    ///     }
+    ///     _ => {}
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// # let n = 0;
+    /// # let a = 3;
+    /// # let b = 4;
+    /// match n {
+    ///     0 if a > b => {
+    ///         return;
+    ///     }
+    ///     _ => {}
+    /// }
+    /// ```
+    #[clippy::version = "1.89.0"]
+    pub DUPLICATE_MATCH_GUARDS,
+    nursery,
+    "a condition in match body duplicating the match guard"
+}
+declare_lint_pass!(DuplicateMatchGuards => [DUPLICATE_MATCH_GUARDS]);
+
+impl<'tcx> LateLintPass<'tcx> for DuplicateMatchGuards {
+    fn check_arm(&mut self, cx: &LateContext<'tcx>, arm: &'tcx Arm<'tcx>) {
+        let Some(guard) = arm.guard else {
+            return;
+        };
+
+        let (arm_body_expr, body_has_block) = if let ExprKind::Block(block, _) = arm.body.kind {
+            if block.stmts.is_empty()
+                && let Some(trailing_expr) = block.expr
```

Comment:
I see this is already making sure that there are no other statements, but we might also want to check that there's [no `#[cfg]`'d out code in between](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/attrs/fn.span_contains_cfg.html). So e.g. I assume the following is currently a false positive:
```rs
match () {
  () if condition {
    #[cfg(...)]
    condition = true;
    if condition {
      
    }
  }
}
```